### PR TITLE
Add `bee-tangle` crate

### PR DESCRIPTION
**Pull request by Adam-Gleave**
# Description of change

Adds the `bee-tangle` crate,

### CODE REVIEW

File: bee-tangle/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+authors = [ "IOTA Stiftung" ]
+edition = "2018"
+description = "Implementation of the IOTA tangle"
+readme = "README.md"
+repository = "https://github.com/iotaledger/bee"
+license = "Apache-2.0"
+keywords = [ "iota", "tangle", "bee", "framework" ]
+homepage = "https://www.iota.org"
+
+[dependencies]
```
Comment (thibault-martinez) [inline bee-tangle/Cargo.toml:13]: Add the default features = false for all deps and dev-deps please

File: bee-tangle/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+# bee-tangle
```
Comment (thibault-martinez) [inline bee-tangle/README.md:1]: Maybe put at least the same description as in the Cargo.toml for now.

File: bee-tangle/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+struct MessageData {
+    message: Arc<Message>,
+    metadata: MessageMetadata,
+}
+
+/// Tangle data structure.
+/// Provides a [`HashMap`] of [`MessageId`]s to [`Message`]s.
+#[derive(Default)]
+pub struct Tangle(Mutex<HashMap<MessageId, MessageData>>);
```
Comment (thibault-martinez) [inline bee-tangle/src/lib.rs]: RwLock instead maybe ?

File: bee-tangle/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /// Retrieves [`MessageMetadata`] from the tangle with a given [`MessageId`].
+    pub async fn get_metadata(&self, message_id: &MessageId) -> Option<MessageMetadata> {
+        self.0.lock().await.get(message_id).map(|data| data.metadata.clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (thibault-martinez) [inline bee-tangle/src/lib.rs]: Can you extract to another file please ?

Comment (Adam-Gleave) [reply bee-tangle/src/lib.rs]: Sure.

File: bee-tangle/src/lib.rs (hunk 3 above)
Comment (Adam-Gleave) [reply bee-tangle/src/lib.rs]: You're right, I think that makes sense.

File: bee-tangle/README.md [hunk 5]
```diff
@@ -0,0 +1,2 @@
...
+Implementation of the IOTA tangle.
```
Comment (thibault-martinez) [inline bee-tangle/README.md:3]:
```suggestion

Implementation of the IOTA tangle.
```

File: bee-tangle/tests/tangle.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use bee_tangle::Tangle;
+use bee_test::rand::message::{metadata::rand_message_metadata, rand_message};
+
+#[tokio::test]
+async fn test_insert() {
```
Comment (thibault-martinez) [inline bee-tangle/tests/tangle.rs]:
```suggestion
async fn insert() {
```

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**